Add `logger` fixture for saving test logs

tbekolay opened the pull request:
And a corresponding py.test `--logs` command line option. See the commit message for more details.

This is a WIP because I get some really weird test failures at unexpected times. E.g., if I run all of the unit tests, `test_rates` fails with
```
>       logging.StreamHandler.__init__(self, stream=TextIO())
E       TypeError: unbound method __init__() must be called with StreamHandler instance as first argument (got CaptureLogHandler instance instead)
```
but if I run this test individually, it runs fine. I'm investigating now, but figured I would make this PR to get opinions on the interface, and whether this is sufficient to replace what the analytics summary data was doing.

hunse commented on nengo/networks/tests/test_actionselection.py:
```diff
@@ -49,7 +49,6 @@ def test_thalamus(Simulator, plt, seed):
-    nengo.log(debug=True)
-    pytest.main([__file__, '-v'])
+    import sys
+    pytest.main(sys.argv)
```
Do we really need these? Can't we just tell people to call `py.test test_whatever.py`?

tbekolay replied:
I'm okay taking them out if that's the consensus. If that's what people want, I'll write up a page in the docs with more details on running unit tests...

hunse replied:
If other people use them, that's fine. I just don't, personally...

jgosmann replied:
me neither

hunse commented on nengo/utils/tests/test_testing.py:
```diff
@@ -59,6 +55,31 @@ def test_analytics_norecord():
+def test_logger_record():
+    logger_obj = Logger('nengo.simulator.logs',
+                        'nengo.utils.tests.test_testing',
+                        'test_logger_record')
```
Could we have a `logger` fixture that returns a logger configured like this (and with parametrized function names, too)?

hunse replied:
And also have it call `logger_obj.__enter__()` and make a callback for exit, à la `plt`?

tbekolay replied:
There is one, this is unit testing the guts of that fixture.

hunse replied:
Ahh, cool, missed that.

hunse commented:
I think this should do most of the stuff that I wanted. It would be cool to add a function to utils that gets the current git information (if you're running in a git repo), so that you can print things like the current branch and commit (hash), since it's easy to forget what timestamps mean.

tbekolay commented:
Okay! I figured out the failed tests; `logging.shutdown()`, which is called in `nengo/utils/test_logging.py`, was causing the fixture to fail in weird ways. I changed those tests, as well as the `nengo.log` function, to hopefully play nicer with other things trying to use loggers.

I also removed all of the `if __name__ == "__main__"` blocks (and made #683 to improve documentation on running unit tests), and made #685 to track @hunse's suggestion.

With that, I think this is ready for srs review.

hunse commented:
I think that addresses all my comments. I scanned through it again and it looks good to me.
